[RFC] core: implement OCALLs

HernanGatta opened the pull request:
# Overview
This PR implements Out Calls, or OCALLs for short. It allows a TA to invoke functions on its CA, effectively `TEEC_InvokeCommand` in reverse.

This is a re-implementation of #3292, but all processing is done in-thread. That is, the normal world thread that invokes a command on a TA, which results in an OCALL, is the one that handles the OCALL, as opposed to using an out-of-thread call queue.

## Known Issues
There is a problem I've run into that I have not been able to figure out. It could very well be my fault, but I'm not sure, given that it reproduces under a specific set of circumstances only. Given the type of problem it is, I'd like to follow up privately in case it turns out to be a security vulnerability. Again, I'm not sure that's what it is, but I'd like to play it safe.

## The life of an OCALL
When opening a session, the CA uses a new API: `TEEC_OpenSessionEx`. This function optionally takes a list of typed settings. One such setting is the OCALL setting. This setting allows the developer to set the OCALL handler function, and an arbitrary context pointer. The function goes on to ascertain that OP-TEE reports OCALL support. If so, the function uses `TEEC_OpenSession` under the covers to open a session. Lastly, it sets the OCALL setting in the session object.

From this point forward, by virtue of having configured OCALLs, the CA's calls into `TEEC_InvokeCommand` are passed via a new IOCTL in the TEE driver called Enclave Call, or ECALL for short. The handler for this IOCTL in the OP-TEE driver initiates a regular function invocation into OP-TEE. If the TA returns in the usual way, that is, it does not request an OCALL, the IOCTL behaves just the same as the Invoke one.

On the other hand, the TA can call `TEE_InvokeCACommand`, whose signature is similar to `TEEC_InvokeTACommand`. This function, implemented in `libutee`, calls into the OCALL PTA.

The OCALL PTA examines the OCALL parameters, and determines if any of them are memory references. If so, it computes their aggregate size and sends an RPC to normal world to request that the CA allocate shared memory of the given size. The driver stack picks up on the request, creates an OCALL context to hold the current call and RPC context, gives it an ID, and the ECALL IOCTL completes.

As part of the IOCTL completion, its values and parameters are now completely changed. A `func` element indicates the function that the TEE Client API must execute, in this case "Allocate shared memory", while the parameters carry the size of the requested buffer. Additionally, the ID of the OCALL context is included for resumption later.

The TEE Client API allocates shared memory the same way the TEE Supplicant does (i.e., if registered memory is available, it is not actually registered with OP-TEE). Once this is done, the result of the operation is wrapped in the IOCTL parameters and a new ECALL IOCTL is issued.

The ECALL IOCTL's logic picks up on the OCALL ID and looks up the OCALL context. The `func` value remains unchanged, so the IOCTL knows this is a reply to the shared memory allocation. After some simple parameter processing, the result is returned to OP-TEE via an RPC resume.

At this stage, the OCALL PTA resumes execution. Its next step is to set up RPC parameters for the OCALL RPC request. The maximum number of RPC parameters is increased to six. As such, the first RPC parameter includes the Command ID for the CA to execute, the second includes the UUID of the TA whence the OCALL request originated, and the remaining four RPC parameters are filled with those of the OCALL, after processing. Processing involves copying values, and copying memory references from TA-mapped memory into the shared memory allocated previously, if necessary. A new RPC is issued.

The RPC is picked up by normal word. The OCALL context is retrieved again, call context is saved, and the ECALL IOCTL's values are updated. The `func` parameter now indicates that the TEE Client API must pass the OCALL request to the CA proper, and the `cmd_id` value stores the Command ID that the CA knows how to interpret (i.e., the OCALL Id). The parameters are those of the OCALL as it originated in the TA, again after proper processing. The ECALL IOCTL completes.

The TEE Client API invokes the OCALL handler set previously via session settings and the CA finally gains control. The handler gets its arbitrary context pointer, the UUID of the requesting TA, and the OCALL parameters. It can modify `INOUT` and `OUTPUT` parameters freely. Once the OCALL handler returns, the TEE Client API processes parameters as needed and issues a new ECALL IOCTL, including the return value of the OCALL and its origin code.

The ECALL IOCTL recognizes this invocation as a resumption by its OCALL ID parameter, and knows that it's a response to an OCALL request by examining the `func` parameter. It performs the appropriate parameter processing and returns to OP-TEE with an RPC resume. The first parameter of the RPC is modified to pass the OCALL return value and origin code.

The OCALL PTA gains control once more. It processes the returned parameters, copying values and buffers as necessary into TA-mapped memory. If shared memory was allocated previously, another round of ECALL IOCTL completion followed by a reply is executed to have the CA free the memory.

All is now done and control is returned to the TA, including the return value and origin code of the OCALL.

Signed-off-by: Hernan Gatta <hegatta@microsoft.com>

HernanGatta commented:
A note on the style fixes commit: I assumed, apparently incorrectly, that commits would be squashed before merging, but that's not the strategy you've all been using. That's my mistake. So, I'll re-write the commit history into something that can be linearly appended to `master` like all the other PRs.

As-is, fixing `checkpath` errors on each commit is going to be a lot of trouble. I've tried rebasing interactively atop `master`, and fix errors on each commit, but that just causes merge conflicts as the rebase progresses, and I'd be fixing errors on code that was ultimately deleted or re-written anyway, so it seems like a waste of time.

Let me know if I got this wrong.

jforissier commented:
@HernanGatta you're right, in the end we expect a set of properly formatted patches, each one addressing a specific need or adding a specific feature.

Coding style fixes can indeed be pain to do during rebase... my suggestion would be to first re-organize your current set of patches using `git rebase -i`. You would squash the things that make sense together and write proper descriptions, but you would not address any coding style issue yet. Then you would rebase again and use the "edit" option to re-format where needed. By doing so you will at least avoid issues with code that is added then removed.

jenswi-linaro commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+	ca_cmd_id = params[0].value.a;
+
+	/* Process OCALL parameters, if any */
+	if (pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) {
```
How about skipping this optimization and instead always require the TA to always supply this?

HernanGatta replied:
Section 4.9.4 of the TEE Internal Core API v1.2.1 spec. states about calling `TEE_InvokeTACommand`:

> Unless all parameter types are set to `TEE_PARAM_TYPE_NONE`, `params` SHALL NOT be `NULL` [...]

I'd like `TEE_InvokeCACommand` to behave as closely to `TEE_InvokeTACommand` as possible. So, `params` can be `NULL` if all parameter types are `NONE`. When this is the case, to remove the check you highlighted, `TEE_InvokeCACommand` would have to pass a dummy array to the PTA.

HernanGatta replied:
Done.

jenswi-linaro replied:
`TEE_InvokeCACommand()` passing a dummy array sounds good.

jenswi-linaro commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+	/* Process OCALL parameters, if any */
+	if (pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) {
+		ca_param_types = params[0].value.b;
+		ca_params = (TEE_Param *)params[1].memref.buffer;
```
A TA may execute in AArch32 while we're at AArch64 here so `TEE_Param` will have a different layout.

HernanGatta replied:
I noticed that the way you've worked around this problem is by turning `TEE_Param` into `utee_param` for transport across the UM/KM boundary. I've taken the same approach. Do you think it's fine that way?

jenswi-linaro replied:
Sounds good.

jenswi-linaro commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		/* Compute memory the CA must allocate for the OCALL params */
+		res = compute_required_mobj_size(ca_param_types, ca_params,
```
This prevents zero copy.
How about letting the TA allocate the buffers first and pass the pointers to those in the `ca_params` above?
That would simplify this function a bit too.

HernanGatta replied:
That's a good point, and I don't disagree, though I'm wondering about the API that would be exposed to TAs if we did that.

Currently, `TEE_Param` has two union'ed structures: one for a `memref` parameter and one for `value` parameter. The former currently behaves like a CA `tmpref`, while the latter behaves like a CA `value`.

What you are suggesting is instead like a CA registered `memref`, where the TA would get an SHM ID of sorts, which the OCALL PTA could then map back to its corresponding `mobj` for transport across worlds. The TA would also have to pass an offset and a size, the same way the CA has to. This would require extending `TEE_Param`.

How had you envisioned this working?

jenswi-linaro replied:
I suppose it would be an extension which would go together with the OCALL extension.
By the way, I think all this should go into the system PTA.

HernanGatta replied:
As far as I can tell, it's not trivial in that what is being suggested is registered memref-type functionality, and I don't think we have enough fields in `TEE_Param`, and associated structures, to make this work, at least not without changing them.

I'm thinking about it...

jenswi-linaro replied:
We can take this after this PR. It's only going to affect the ABI between TA and OP-TEE, so we can let that change with regards to OCALL for a while.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -547,3 +547,6 @@ CFG_SHOW_CONF_ON_BOOT ?= n
 CFG_CORE_TPM_EVENT_LOG ?= n
+
+# Enables support for OCALLs
+CFG_OCALL ?= y
```
I don't think `_CA_` would add anything useful. The name of the feature is OCALL.
The comment describing the config could mention that this is an RPC into the Client Application.

HernanGatta replied:
Added more info.

jforissier commented on lib/libutee/tee_api.c:
```diff
@@ -10,10 +10,17 @@
 #include <types_ext.h>
 #include <user_ta_header.h>
 #include <utee_syscalls.h>
+#if defined(CFG_OCALL)
```
`#if` not necessary I think

HernanGatta replied:
Removed.

jforissier commented on lib/libutee/tee_api.c:
```diff
@@ -10,10 +10,17 @@
 static const void *tee_api_instance_data;
 
+#if defined(CFG_OCALL)
+TEE_TASessionHandle tee_api_ocall_session;
```
`static`?

HernanGatta replied:
This is also linked against in `user_ta_entry.c` to tear down the session after the TA exits.

jforissier replied:
Right.

jforissier commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -255,6 +260,11 @@ static TEE_Result entry_close_session(unsigned long session_id)
 	TA_CloseSessionEntryPoint(session->session_ctx);
 
+#if defined(CFG_OCALL)
+	if (tee_api_ocall_session)
```
Could we avoid the `#if` (here and above) with the following? (untested):
```C
	if (IS_ENABLED(CFG_OCALL) && tee_api_ocall_session)
		...
```

HernanGatta replied:
We could do this, but then we'd have to always declare `tee_api_ocall_session`, regardless of the value of `CFG_OCALL`. That is, if OCALL support is disabled, that variable would still exist. I would imagine that the linker would realize it's untouched and discard it.

Preference?

jforissier replied:
> I would imagine that the linker would realize it's untouched and discard it.

Indeed, because we're using `-fdata-sections` and `--gc-sections`.
I think we should try to use `IS_ENABLED()` as much as possible, because it avoids `#ifdef` which is good for legibility and also build coverage in CI.

HernanGatta replied:
Done.

Would you like to see `IS_ENABLED` for [capability reporting](https://github.com/OP-TEE/optee_os/pull/3673/files#diff-7a16795083f7db7775a0c1cf31782e9f), too?

We have there another case of a different existing style, in this case not using the macro.

jforissier commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+	struct tee_ta_session *s;
+	TEE_Result res;
+
+	(void)session_ctx;
```
Add `__unused` to the parameter instead

HernanGatta replied:
Added.

HernanGatta commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -70,6 +70,7 @@
 #define TEE_ORIGIN_COMMS                0x00000002
 #define TEE_ORIGIN_TEE                  0x00000003
 #define TEE_ORIGIN_TRUSTED_APP          0x00000004
+#define TEE_ORIGIN_CLIENT_APP           0x00000005
```
I'd like to mirror the naming convention that's already established. `TEE_ORIGIN_OCALL_APP` would suggest an application that can receive OCALLs. By that token, a TA is an application that can receive function invocations, so `TEE_ORIGIN_TRUSTED_APP` would be `TEE_ORIGIN_INVOKE_FUNCTION_APP`, which is not what's used already. Additionally, "Client Application" is GP terminology, so I'd prefer we stick to that.

etienne-lms replied:
I understand. Yet, this origin ID is meaningful to the TA that does not OCall. Will it be meaningful for the CA that originate the TA invocation?

HernanGatta replied:
Is your concern that seeing as this origin ID would only be used in the context of an OCALL, failing to attach the name "OCALL" to this origin ID would therefore entail that it ought to also be meaningful to TAs that do not use OCALLs?

I would counter with the inverse of your argument in that this origin ID would be seen only by TAs that perform OCALLs. Hence, TAs that do not perform OCALLs need not concern themselves with it.

Additionally, it seems to me that you intend this origin ID to mean "this error originated form the client application while it attempted to handle an OCALL". To that, I would suggest that the only way for an error to originate form the client application is for it to have arisen while the latter was handling an OCALL, hence attaching "OCALL" to this origin ID is redundant.

etienne-lms replied:
resolved

HernanGatta commented on lib/libutee/include/pta_ocall.h:
```diff
@@ -0,0 +1,11 @@
+		{ 0x05ece642, 0x903f, 0x47bf, \
+			{ 0x94, 0xeb, 0x07, 0xd6, 0x14, 0x61, 0x73, 0xa1 } }
+
+#define PTA_OCALL_SEND      0
```
Good point.

HernanGatta replied:
Fixed.

HernanGatta commented on core/include/optee_rpc_cmd.h:
```diff
@@ -148,6 +150,20 @@
 #define OPTEE_RPC_CMD_BENCH_REG		20
 
+/*
+ * Send a command to the Client Application.
```
Agreed.

HernanGatta replied:
Fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,322 @@
+	return TEE_SUCCESS;
+}
+
+#define MEMREF_CHECK_AND_SET_RPC(x) 					       \
```
Indeed, will simplify; this is ugly.

HernanGatta replied:
Fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+		case TEE_PARAM_TYPE_NONE:
+			rpc_params[n].attr = THREAD_PARAM_ATTR_NONE;
+			break;
+		case TEE_PARAM_TYPE_VALUE_INPUT:
```
Ah, I like it; I wanted to get rid of the duplication but couldn't really due to the macro. I had seen code like what you suggested elsewhere in OP-TEE, but I figured that since the macro to set thread parameters was there, one is meant to use them.

Thank you for the suggestion.

HernanGatta replied:
Added as suggested.

etienne-lms commented on lib/libutee/include/tee_internal_api_extensions.h:
```diff
@@ -34,6 +34,17 @@ TEE_Result TEE_CacheClean(char *buf, size_t len);
+ * The semantics are identical to TEEC_InvokeCommand, except that memory
+ * references with size zero are not supported, in the opposite direction.
+ */
+TEE_Result TEE_InvokeCACommand(uint32_t cancellationRequestTimeout,
```
Sorry nitpicking on labels, I think it would be clearer if this is renamed `TEE_InvokeOCall()` or `TEE_InvokeOCallCommand()`

HernanGatta replied:
Nothing to be sorry about, I welcome feedback, it makes me think about my silly ideas in different ways.

I'm not opposed to the change. The reason I chose `TEE_InvokeCACommand` is because it neatly mirrors `TEE_InvokeTACommand`, but `TEE_InvokeOcall` would also work.

@jenswi-linaro, what do you say?

jenswi-linaro replied:
I liked `TEE_InvokeCACommand()` too because it neatly mirrors `TEE_InvokeTACommand()`. I don't have a strong opinion on this.

jbech-linaro replied:
+1 for `TEE_InvokeCACommand()`.

etienne-lms commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+					     TEE_Param params[TEE_NUM_PARAMS],
+					     size_t *required_size)
+{
+	TEE_Param *param;
```
generic comment: always initialize local variables where defined. See [coding-standards](https://optee.readthedocs.io/en/latest/general/coding_standards.html#coding-standards), 3rd bullet.

HernanGatta replied:
Interesting. This requirement cuts both ways though: always initializing variables precludes the compiler from warning the developer when the variable is used in a location where its value should have been updated prior to the usage. I've run into a few of those in fact while writing this patch set.

From my days of working on Windows, we'd leave uninitialized variables which are expected to be filled later on. That way, if our control flow was busted and a variable was used without having been set to the expected value, the compiler would bark at us. And we compile with warnings as errors, so we'd fail the build altogether.

I'll follow the coding standards, but with reservations :) .

jbech-linaro replied:
The rationale is that we want all variables to have a well defined value, i.e., to have the developer think about it and come up with a good value instead of having what often is a random value otherwise (static's etc are an exception). The reason why we've introduced this is because we've been bitten by a couple of security issues in the past due to uninitialized variables. As anything else, there are pros and cons.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,344 @@
+							   ca_param->value.b,
+							   0);
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
```
Added as suggested.

HernanGatta commented:
@jforissier

> in the end we expect a set of properly formatted patches

All fixed up, except for this:
```
CHECK: Prefer using the BIT macro
#25: FILE: core/arch/arm/include/sm/optee_smc.h:280:
+#define OPTEE_SMC_SEC_CAP_OCALL			(1 << 4)
```

I could easily change that to `BIT(4)`, but that wouldn't follow the style of the rest of the file. Would you like me to fix up the rest of the file with `BIT(x)`, or do we ignore this complaint?

jforissier commented:
> @jforissier
> 
> > in the end we expect a set of properly formatted patches
> 
> All fixed up, except for this:
> 
> ```
> CHECK: Prefer using the BIT macro
> #25: FILE: core/arch/arm/include/sm/optee_smc.h:280:
> +#define OPTEE_SMC_SEC_CAP_OCALL			(1 << 4)
> ```
> 
> I could easily change that to `BIT(4)`, but that wouldn't follow the style of the rest of the file.
> 
> Would you like me to fix up the rest of the file with `BIT(x)`, or do we ignore this complaint?

I'd rather fix that too since I see no reason not to ;-) Please introduce a prior commit changing the existing defines.

jbech-linaro commented:
> > @jforissier
> > > in the end we expect a set of properly formatted patches
> > 
> > 
> > All fixed up, except for this:
> > ```
> > CHECK: Prefer using the BIT macro
> > #25: FILE: core/arch/arm/include/sm/optee_smc.h:280:
> > +#define OPTEE_SMC_SEC_CAP_OCALL			(1 << 4)
> > ```
> > 
> > 
> > I could easily change that to `BIT(4)`, but that wouldn't follow the style of the rest of the file.
> > Would you like me to fix up the rest of the file with `BIT(x)`, or do we ignore this complaint?
> 
> I'd rather fix that too since I see no reason not to ;-) Please introduce a prior commit changing the existing defines.

Also, please change it to use "5" instead of "4", since we already have other pull requests (NULL parameter support patches) in the pipe where "4" is used.

HernanGatta commented:
> The rationale is [..]

Thank for you the explanation. Seeing the security rationale in the coding standards does certainly make sense, one wouldn't want to leak stack data inadvertently.

> I'd rather fix that too since I see no reason not to ;-) Please introduce a prior commit changing the existing defines.

Done.

> Also, please change it to use "5" instead of "4", since we already have other pull requests (NULL parameter support patches) in the pipe where "4" is used.

Good call, done.

HernanGatta commented:
All green, cleaning up the [Linux PR](https://github.com/linaro-swg/linux/pull/72) next.

jenswi-linaro commented:
> 1.1. While I'm not arguing for the following, I wonder: would anyone suggest turning this functionality into a system call seeing as TEE_InvokeTACommand is backed by one?

We don't add new syscalls unless there's no other practical way.

> 3. Zero-copy now or later?
> 3. We could do it later for no other reason than to get the current changes moving.
> 3.1 Can anyone think of any risks regarding, say, backwards compatibility, or otherwise, if we wait?

There is the backwards compatibility issue with the ABI change, the API will remain the same the TA just needs to be recompiled or at least relinked.

> 4. I'd like to follow up as suggested about the known issue in the PR description.

Please contact me privately, jens.wiklander@linaro.org and I'll take a look.

HernanGatta commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -264,20 +264,20 @@
  */
 /* Normal world works as a uniprocessor system */
-#define OPTEE_SMC_NSEC_CAP_UNIPROCESSOR		(1 << 0)
+#define OPTEE_SMC_NSEC_CAP_UNIPROCESSOR		BIT(0)
```
Added.

HernanGatta commented on lib/libutils/ext/include/compiler.h:
```diff
@@ -70,6 +70,12 @@
 		       __GNUC_PATCHLEVEL__)
 
+#if __GCC_VERSION >= 70100
+#define fallthrough	__attribute__((fallthrough));
```
Removed.

HernanGatta commented:
Thank you for your review; I've applied the tags to the commits.

etienne-lms commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+	if (mobj)
+		thread_rpc_free_client_app_payload(mobj);
+
+	return TEE_SUCCESS;
```
`return res;`?

HernanGatta replied:
Yes, that was another one of the fixes I had in the wrong commit. It's patched up now.

etienne-lms commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+
+	res = thread_rpc_cmd(OPTEE_RPC_CMD_OCALL, rpc_num_params, rpc_params);
+	if (res != TEE_SUCCESS) {
+		mobj_put(mobj);
```
why `mobj_put()` here and `thread_rpc_free_client_app_payload()` on exit?

HernanGatta replied:
I added a comment to explain the rationale. Do you see a different possible behavior?

HernanGatta replied:
Modified as suggested.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return tee_mmu_check_access_rights(uctx, flags, (uaddr_t)buffer, size);
```
Fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+
+	struct mobj *mobj = NULL;
+	size_t mobj_size = 0;
+	void *mobj_va = 0;
```
Fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+	size_t mobj_size = 0;
+	void *mobj_va = 0;
+
+	struct thread_param rpc_params[THREAD_RPC_MAX_NUM_PARAMS] = { 0 };
```
Fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+		goto exit;
+
+	params[0].value.a = rpc_params[0].u.value.b;  // OCALL ret val
+	params[0].value.b = rpc_params[0].u.value.c;  // OCALL ret val origin
```
Replaced.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+		return ocall_send(session, calling_session, param_types,
+				  params);
+	default:
+		break;
```
Done.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,125 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+		}
+	}
+
+	/* Convert the OCALL's parameter into a utee_params structure */
```
Fixed.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,125 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+	uint32_t ret_origin = TEE_ORIGIN_TEE;
+
+	struct utee_params pta_up = { 0 };
+	struct utee_params ocall_up = { 0 };
```
Fixed.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,125 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+			break;
+		default:
+			ocall_up.vals[n * 2] = 0;
+			ocall_up.vals[n * 2 + 1] = 0;
```
Removed.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,125 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+	    ret_origin == TEE_ORIGIN_API)
+		return res;
+
+	if (res != TEE_SUCCESS &&
```
How about now? The logic I pushed mirrors that in `Invoke_TACommand`.

HernanGatta commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -84,6 +85,8 @@ static void uninit_instance(void)
 {
 	__utee_gprof_fini();
 	TA_DestroyEntryPoint();
+	if (IS_ENABLED(CFG_OCALL) && tee_api_ocall_session)
```
Fixed.

HernanGatta commented on core/include/optee_rpc_cmd.h:
```diff
@@ -148,6 +150,20 @@
+ * [in]     value[1].a	    UUID of TA whence OCALL originated (HI bits)
+ * [out]    value[1].b	    UUID of TA whence OCALL originated (LO bits)
+ * [-]      value[1].c	    Unused
+ *
```
Done.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+
+	ocall_id = params[0].value.a;
+	ocall_pt = params[0].value.b;
+	ocall_up = (struct utee_params *)params[1].memref.buffer;  // User ptr
```
Replaced.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+	ocall_up = (struct utee_params *)params[1].memref.buffer;  // User ptr
+
+	res = check_access_rights(&utc->uctx, ocall_up);
+	if (res != TEE_SUCCESS)
```
Changed all tests as requested.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+
+	res = post_process(rpc_params + 2, ocall_up, mobj_va);
+	if (res != TEE_SUCCESS)
+		goto exit;
```
That was fixed by the changes I had in the wrong commit; it's OK now.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+}
+
+pseudo_ta_register(.uuid = PTA_UUID, .name = PTA_NAME,
+		   .flags = PTA_DEFAULT_FLAGS,
```
That was added in the changes I had in the wrong commit; it's there now.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -330,7 +330,7 @@ static TEE_Result ocall_send(struct tee_ta_session *session,
 	if (mobj)
 		thread_rpc_free_client_app_payload(mobj);
 
-	return TEE_SUCCESS;
```
Sorry about that; fixed.

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+ * params[2..5]: < OCALL Params, if any >
+ *
+ * There are two return codes:
+ *   res: Did sending the RPC work?
```
I added more details to the comment. Do you think it's clearer?

HernanGatta commented on core/pta/ocall.c:
```diff
@@ -0,0 +1,364 @@
+ * params[0]: VALUE IN/OUT
+ *   IN: < HOST CMD ID,  Param Types   >
+ *  OUT: < CMD Ret Val, CMD Ret Origin >
+ * params[1]: < [INOUT] TEE_Param[] >
```
Thank you for your suggestions. I've rearranged the comment following your idea.

I also removed the OCALL param types from the TA interface; they aren't needed there since they are already included in `struct utee_params`.

HernanGatta commented:
Thank you for your review, Etienne. I've addressed your feedback and applied the tag to commit "core: OCALL capability".

HernanGatta commented on lib/libutee/include/pta_ocall.h:
```diff
@@ -0,0 +1,30 @@
+ * itself, whose interpretation is up to the CA & TA.
+ */
+#define PTA_OCALL_SEND			0
+
```
All the other PTA headers have an extra line between the last command ID and the bottom `#endif`. I don't really mind either way, I'd just like to stay consistent. I'll still remove the line if you'd like, though.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,124 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+	if (returnOrigin)
+		*returnOrigin = ret_origin;
+
+	if (ret_origin == TEE_ORIGIN_TEE &&
+	    (res != TEE_SUCCESS ||
+	     res != TEE_ERROR_OUT_OF_MEMORY ||
+	     res != TEE_ERROR_TARGET_DEAD))
+		TEE_Panic(res);
```
@etienne-lms: This little tidbit is back. I realized that if we panic on things like bad parameters regardless of the error origin, then the few tests we perform above in this function would cause a panic. It seems to me that'd be too harsh.

Do you see any security concerns regarding this more permissive logic (in comparison with `TEE_InvokeTACommand`)?

etienne-lms replied:
I'll look at that.

HernanGatta replied:
Maybe we can return early without panicking from anywhere above the `exit` label, except for an error in the return code of `utee_invoke_ta_command`; then, we'd jump to the `exit` label and perhaps decide to panic.

HernanGatta commented on lib/libutee/include/pta_ocall.h:
```diff
@@ -0,0 +1,30 @@
+ * which will receive the OCALL request.
+ *
+ * [in/out] value[0].a:	Host OCALL ID (IN), OCALL return value (OUT)
+ * [in]     value[0].b:	OCALL parameter types (IN), OCALL error origin (OUT)
```
Fixed (and updated with the removal of the OCALL parameter types parameter).

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -231,6 +234,120 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+
+	size_t n = 0;
+
+	/* Open TA2TA session with the OCALL PTA, if necessary */
```
Fixed.

HernanGatta commented:
@etienne-lms: Jens has expressed a preference to have the code in the OCALL PTA moved to the System PTA. Do you have a preference?

etienne-lms commented:
> @etienne-lms: Jens has expressed a preference to have the code in the OCALL PTA moved to the System PTA. Do you have a preference?

Fine either ways.

HernanGatta commented:
I've moved OCALL support into the System PTA.

**Question**
1. I had made OCALL support optional because it's not part of the GP specs.;
2. The System PTA is already optional for, at least in part, that same reason according to [the documentation](https://github.com/OP-TEE/optee_os/blob/ef18a901f5efd996576d917bee54d9797687a12f/mk/config.mk#L476) that describes the `CFG_SYSTEM_PTA` option;
3. I've kept the ability to switch OCALL support off even after moving it to the System PTA.

Do we want that, or do we just bake OCALL support into OP-TEE if the developer chooses to build the System PTA? I can argue it from both sides and don't find any side more compelling than the other.

jforissier commented:
> Do we want that, or do we just bake OCALL support into OP-TEE if the developer chooses to build the System PTA? I can argue it from both sides and don't find any side more compelling than the other.

As far as I'm concerned I'm happy with the approach you chose. People with constrained memory resources might want to disable some features to save space, and extensions like this are certainly good candidates (although I did not check how much space this adds to the binary).

HernanGatta commented:
> although I did not check how much space this adds to the binary

I'll check. We can save even more space by `#ifdef`ing out `TEE_InvokeCACommand` and just returning not implemented.

etienne-lms commented:
Can one remove the Stale label from this P-R?

jforissier commented:
> Can one remove the Stale label from this P-R?

Done.

jforissier commented:
Why not fold "core: OCALL capability" into "core: OCALL support"? It seems a bit artifical to split them IMO.

HernanGatta commented:
> Why not fold "core: OCALL capability" into "core: OCALL support"? It seems a bit artifical to split them IMO.

That's fair, folded as suggested. Note that this dropped @etienne-lms' `Reviewed-by` tag on `core: OCALL capability`.

> > although I did not check how much space this adds to the binary
> 
> I'll check. We can save even more space by `#ifdef`ing out `TEE_InvokeCACommand` and just returning not implemented when OCALLs are disabled.

In the latest changes, both the system PTA OCALL functionality as well as `TEE_InvokeCACommand` are `#ifdef`ed out and return `TEE_ERROR_NOT_IMPLEMENTED` when `CFG_OCALL=n`.

With `OPTEE_OS_COMMON_FLAGS += CFG_ARM64_core=y DEBUG=n` in `qemu_v8.mk`, `CFG_OCALL=y` adds 992 bytes to the `.text` section of `tee.elf`. `tee.bin` does not change size regardless of the setting, owing, I would surmise, to padding.

jforissier commented on lib/libutee/include/pta_system.h:
```diff
@@ -189,4 +189,21 @@
+ * Send an OCALL to the calling TA's Client Application
+ *
+ * Note that if TA 1 invokes TA 2, which invokes TA N+1, it is the CA of TA 1
+ * which will receive the OCALL request.
```
A minor thing. While reviewing I started to wonder if this is really a good thing. I thought about possible conflicts with command IDs and CAs executing commands they are not supposed to execute as they come from a TA they have not invoked directly. But then I realized that the TA UUID is passed to the CA so the context is clear.

So, how about reminding the reader that in any case the OCALL requests contains the UUID of the originator TA?

HernanGatta replied:
I have added a comment to that effect.

jforissier commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -70,6 +70,7 @@
 #define TEE_ORIGIN_COMMS                0x00000002
 #define TEE_ORIGIN_TEE                  0x00000003
 #define TEE_ORIGIN_TRUSTED_APP          0x00000004
+#define TEE_ORIGIN_CLIENT_APP           0x00010000
```
This falls in the GlobalPlatform reserved range (TEE Internal Core API v1.2.1 section 4.2.3). Please use `0xF0000001` instead.
Also `lib/libutee/include/tee_api_defines_extensions.h` would be a better place.

HernanGatta replied:
Thank you for catching that! Changed and moved.

jforissier commented on lib/libutee/include/tee_internal_api_extensions.h:
```diff
@@ -34,6 +34,17 @@ TEE_Result TEE_CacheClean(char *buf, size_t len);
+/*
+ * Send an OCALL to the Client Application
+ *
+ * The semantics are identical to TEEC_InvokeCommand. but in the opposite
```
`to TEEC_InvokeCommand() but`

HernanGatta replied:
Fixed.

jforissier commented on lib/libutee/tee_api_private.h:
```diff
@@ -8,6 +8,8 @@
 #include <utee_types.h>
 
+/* From tee_api.c */
+extern TEE_TASessionHandle tee_api_system_session;
```
I guess that should have double underscores (`__tee_api_system_session`) because it is a global symbol for internal use and we don't want any clash with user-defined stuff.

@jenswi-linaro there are a few instances of this rule not being enforced, such as the `utee_` syscall wrappers. Should they be renamed? (that would be taken separately of this PR of course).

jenswi-linaro replied:
@jforissier, yes those could be renamed. I guess there's quite a bit of that in other corners of libutee too, one small step at a time is the way to go here I think.

HernanGatta replied:
Renamed.

jforissier commented:
@HernanGatta do you think you can spend some more time addressing the last comments on this series? AFAICT there is not much left to do before this can be merged (and same for the other PRs: libteec, xtest and the linux driver). It would be really nice to have this in the next release (3.10.0) which is [planned](https://optee.readthedocs.io/en/latest/general/releases.html#release-dates) for August 21st.

HernanGatta commented:
> @HernanGatta do you think you can spend some more time addressing the last comments on this series? AFAICT there is not much left to do before this can be merged (and same for the other PRs: libteec, xtest and the linux driver). It would be really nice to have this in the next release (3.10.0) which is [planned](https://optee.readthedocs.io/en/latest/general/releases.html#release-dates) for August 21st.

Unfortunately, the timing on my end didn't line up. Hopefully we can get this in by 3.12.0 in January, though.

I've addressed your comments. An additional change is that I've set `CFG_OCALL=n` if `CFG_CORE_SEL1_SPMC=y`. I'm not familiar with FF-A and the corresponding build did not pass seeing as the RPC functions for memory allocation and free are reimplemented for that case in a different way.

I presume there should be a way to still support OCALLs on FF-A, though. Is it possible to emulate FF-A with QEMU?

jenswi-linaro commented:
Disabling OCALL now if FF-A is in use is OK in my opinion. FF-A is still in an experimental state. Adding OCALL support with FF-A later on when things has settled should not be hard.

jforissier reviewed (commented):
LGTM, one small concern below.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -245,6 +248,136 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+	size_t n = 0;
+
+	/* Open session with the System PTA, if necessary */
+	if (!__tee_api_system_session) {
```
Very true. I modified `invoke_system_pta` to use this global variable instead.

After pushing I had another idea to clean this up, a few more changes incoming...

HernanGatta replied:
No, the idea I had is uglier because it requires defining `tee_ocall` with the exact same parameters as `TEE_InvokeCACommand` in the exact same header file to avoid exposing `struct utee_params` via `tee_internal_api_extensions.h`.

jforissier replied:
> Very true. I modified `invoke_system_pta` to use this global variable instead.

OK, looks reasonable.

HernanGatta commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -165,6 +165,8 @@ static void uninit_instance(void)
 	__utee_gprof_fini();
 	TA_DestroyEntryPoint();
+	if (__tee_api_system_session)
+		TEE_CloseTASession(__tee_api_system_session);
```
Removed.

HernanGatta commented on lib/libutee/tee_api.c:
```diff
@@ -245,6 +248,136 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+				      cancellationRequestTimeout,
+				      PTA_SYSTEM_OCALL,
+				      &pta_up,
+				      &ret_origin);
```
I considered funneling this call via a function in `lib/libutee/tee_system_pta.c` which would eventually land in `invoke_system_pta` and reuse the static `sess` variable to manage the session to the system PTA.

However, this function calls `TEE_InvokeTACommand`, which performs parameter processing and creates copies of any `TEE_PARAM_TYPE_MEMREF_*` buffers to isolate the caller TA's memory for the TA being called. This is unnecessary for OCALLs since the callee is the system PTA which can read the caller's memory anyway. To avoid this copy, I construct the `struct utee_params` for the call manually.

If I were to create a wrapper for this call in `lib/libutee/tee_system_pta.c`, I'd have to either create the call's `struct utee_params` in `Invoke_CACommand` and pass that to the wrapper, which would require exposing `struct utee_params` in `tee_internal_api_extensions.h`, or move the entire logic of `Invoke_CACommand` to `lib/libutee/tee_system_pta.c`. This would yield a new function with the same signature as `Invoke_CACommand`, which doesn't seem clean to me.

jforissier replied:
I see, so it makes sense to keep most of the processing here.

jforissier commented:
@HernanGatta ping?

etienne-lms commented:
Hello @HernanGatta and all,
Some info on this P-R series.
I am working on SCMI support in OP-TEE where OP-TEE Core exposes system resources to REE kernel: clocks, sensors, reset controller, and few more. One can have a look at the [SCMI specification](https://developer.arm.com/documentation/den0056/latest). 
Using OCALLs, we could provision an OP-TEE thread for SCMI message exchange between REE kernel and TEE core. I plan to create some P-Rs for this based on your work, using only the REE/TEE kernels parts and maybe not the SHM parts, and get one's feedback.
Regards

jforissier commented:
@etienne-lms this PR was discussed in the last LOC meeting ([link](https://docs.google.com/document/d/15XsqgGktCrRRWiqyaz-erp_cZykwGjkBkhMD2XtUlUY/edit#)) and everyone agreed that it makes sense to revive this PR, so your contributions are welcome :) As mentioned in the notes, the difficult part may be related to some reference counting issue in the Linux driver patches so we concluded it would be good to start by clarifying this point.

etienne-lms commented:
Thanks for the link :)

HernanGatta commented:
Hello all, my apologies for the hiatus. I'm back and looking at this again, particularly the issues that were brought up regarding memory references. I'll report back with what I found. I'm glad to see functionality being implemented atop OCALLs :) .

HernanGatta commented:
Latest push is merely a rebase. Jens' comments still to be addressed.
